server/clustering.go

```diff
@@ -513,40 +533,42 @@ func (r *raftFSM) Apply(l *raft.Log) interface{} {
 	switch op.OpType {
 	case spb.RaftOperation_Publish:
 		// Message replication.
-		var (
-			c   *channel
-			err error
-		)
-		for _, msg := range op.PublishBatch.Messages {
-			// This is a batch for a given channel, so lookup channel once.
-			if c == nil {
-				c, err = s.lookupOrCreateChannel(msg.Subject)
-				// That should not be the case, but if it happens,
-				// just bail out.
-				if err == ErrChanDelInProgress {
-					return nil
-				} else if err == nil && !c.lSeqChecked {
-					// If msg.Sequence is > 1, then make sure we have no gap.
-					if msg.Sequence > 1 {
-						// We pass `1` for the `first` sequence. The function we call
-						// will do the right thing when it comes to restore possible
-						// missing messages.
-						err = s.raft.fsm.restoreMsgsFromSnapshot(c, 1, msg.Sequence-1, true)
-					}
-					if err == nil {
-						c.lSeqChecked = true
-					}
+		if len(op.PublishBatch.Messages) == 0 {
+			return nil
+		}
+		// This is a batch for a given channel, so lookup channel once.
+		msg := op.PublishBatch.Messages[0]
+		c, err := r.lookupOrCreateChannel(msg.Subject, op.ChannelID)
+		if err != nil {
+			goto FATAL_ERROR
+		}
+		// `c` will be nil if the existing channel has an ID > than op.ChannelID.
+		// This will be the case in RAFT log replay if we have several "versions"
+		// of the same channel. In that case, simply ignore the log replay.
+		if c == nil {
+			return nil
+		}
+		if !c.lSeqChecked {
+			// If msg.Sequence is > 1, then make sure we have no gap.
+			if msg.Sequence > 1 {
+				// We pass `1` for the `first` sequence. The function we call
+				// will do the right thing when it comes to restore possible
+				// missing messages.
+				if err = s.raft.fsm.restoreMsgsFromSnapshot(c, 1, msg.Sequence-1, true); err != nil {
+					goto FATAL_ERROR
 				}
 			}
-			if err == nil {
-				_, err = c.store.Msgs.Store(msg)
-			}
-			if err != nil {
-				panic(fmt.Errorf("failed to store replicated message %d on channel %s: %v",
-					msg.Sequence, msg.Subject, err))
+			c.lSeqChecked = true
+		}
+		for _, msg = range op.PublishBatch.Messages {
+			if _, err = c.store.Msgs.Store(msg); err != nil {
+				goto FATAL_ERROR
 			}
 		}
 		return c.store.Msgs.Flush()
+	FATAL_ERROR:
+		panic(fmt.Errorf("failed to store replicated message %d on channel %s: %v",
```

Comment:
No way to recover from this?

Reply:
No. The RAFT log is committed at this point, if we fail to add to our message store we would take the risk of being out of sync and add later a higher sequence which will create gaps.

Reply:
How would an operator restore a working system after this failure?

Reply:
The failure is due to not be able to write into the message store (disk full, storage failure)? So making some room or if the disk is toasted, the node would be restarted without any state and will be reconstructed from the running leader.

Reply:
Thx